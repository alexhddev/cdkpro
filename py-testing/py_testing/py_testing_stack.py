from aws_cdk import (
    Stack,
    aws_lambda,
    aws_s3    
)
from constructs import Construct

class PySimpleStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        cool_lambda = aws_lambda.Function(self, 'SimpleLambda',
            runtime=aws_lambda.Runtime.PYTHON_3_11,
            handler='index.handler',
            code=aws_lambda.Code.from_inline('print()')
        )
        bucket = aws_s3.Bucket(self, 'SimpleBucket',
            versioned=True
        )
        bucket.grant_read(cool_lambda)

