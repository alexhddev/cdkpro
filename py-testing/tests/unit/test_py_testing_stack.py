import aws_cdk as core
import aws_cdk.assertions as assertions
import pytest
from aws_cdk.assertions import Match, Capture

from py_testing.py_testing_stack import PySimpleStack


@pytest.fixture(scope="session")
def simple_template():
    app = core.App()
    stack = PySimpleStack(app, "py-testing")
    template = assertions.Template.from_stack(stack)
    return template


def test_lambda_props(simple_template):
    simple_template.has_resource_properties(
        "AWS::Lambda::Function", {"Runtime": "python3.11"}
    )
    simple_template.resource_count_is("AWS::Lambda::Function", 1)


def test_lambda_runtime_with_matcher(simple_template):
    simple_template.has_resource_properties(
        "AWS::Lambda::Function", {"Runtime": Match.string_like_regexp("python")}
    )


def test_lambda_bucket_with_matchers(simple_template):
    simple_template.has_resource_properties(
        "AWS::IAM::Policy",
        Match.object_like(
            {
                "PolicyDocument": {
                    "Statement": [
                        {
                            "Resource": [
                                {
                                    "Fn::GetAtt": [
                                        Match.string_like_regexp("SimpleBucket"),
                                        "Arn",
                                    ]
                                },
                                Match.any_value(),
                            ]
                        }
                    ]
                }
            }
        ),
    )


def test_lambda_actions_with_captors(simple_template):
    lambda_actions_captor = Capture()
    simple_template.has_resource_properties(
        "AWS::IAM::Policy",
        {"PolicyDocument": {"Statement": [{"Action": lambda_actions_captor}]}},
    )

    expected_actions = [
        "s3:GetBucket*",
        "s3:GetObject*",
        "s3:List*"
       ]
    
    assert sorted(lambda_actions_captor.as_array()) == sorted(expected_actions)
