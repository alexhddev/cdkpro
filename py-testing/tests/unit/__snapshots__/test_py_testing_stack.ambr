# serializer version: 1
# name: test_bucket_props_with_snapshot
  dict({
    'SimpleBucket250AC437': dict({
      'DeletionPolicy': 'Retain',
      'Properties': dict({
        'VersioningConfiguration': dict({
          'Status': 'Enabled',
        }),
      }),
      'Type': 'AWS::S3::Bucket',
      'UpdateReplacePolicy': 'Retain',
    }),
  })
# ---
